### PR TITLE
chore: reverse MemoryOverlayStateProvider block order

### LINKED ISSUE #9825: reverse MemoryOverlayStateProvider block order
Currently the MemoryOverlayStateProvider expects the list of blocks to be `[old -> new]`

https://github.com/paradigmxyz/reth/blob/c4bf5bb8ca271a6055f5af8431fd0bfb4ba79feb/crates/chain-state/src/memory_overlay.rs#L13-L14

https://github.com/paradigmxyz/reth/blob/c4bf5bb8ca271a6055f5af8431fd0bfb4ba79feb/crates/chain-state/src/memory_overlay.rs#L34-L34

but this is usually populated by walking backwards:

https://github.com/paradigmxyz/reth/blob/c4bf5bb8ca271a6055f5af8431fd0bfb4ba79feb/crates/engine/tree/src/tree.rs#L732-L735

so it would be more efficient if we'd do [new -> old] so appending a parent block is a simple push

### PR DESCRIPTION
**Pull request by fgimenez**
Closes: #9825

### CODE REVIEW

File: crates/chain-state/src/in_memory.rs [hunk 1]
```diff
@@ -455,24 +455,24 @@ impl BlockState {
...
     }
 
     /// Returns a vector of `BlockStates` representing the entire in memory chain,
-    /// including self as the last element.
+    /// including self as the first element.
     pub fn chain(&self) -> Vec<&Self> {
         let mut chain = self.parent_state_chain();
-        chain.push(self);
+        chain.insert(0, self);
         chain
```
Comment (mattsse) [inline crates/chain-state/src/in_memory.rs:476]: one improvement we can make here or in a followup is adding another function that is 

`fn append_parent_chain(&self, chain: &mut Vec)` so we don't need to insert0 here which needs to shift the entire chain

File: crates/chain-state/src/memory_overlay.rs
Comment (mattsse) [inline crates/chain-state/src/memory_overlay.rs:1]: can we document the expected order on the new function and the field of MemoryOverlayStateProvider

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, doc nit and suggestion

File: crates/chain-state/src/memory_overlay.rs
Comment (fgimenez) [reply crates/chain-state/src/memory_overlay.rs:1]: sure done here https://github.com/paradigmxyz/reth/pull/9826/commits/7236f14b188179b059bee576508ad7f9313fb0a4

File: crates/chain-state/src/in_memory.rs (hunk 1 above)
Comment (fgimenez) [reply crates/chain-state/src/in_memory.rs]: added here https://github.com/paradigmxyz/reth/pull/9826/commits/7236f14b188179b059bee576508ad7f9313fb0a4 ptal

**Review by mattsse — APPROVED** [review]

**Pull request merged**